### PR TITLE
fix(api): make nil value in nvim_set_option_value clear local value

### PR DESCRIPTION
**Pull request by gpanders**
Fix a bug in `nvim_set_option_value` that would set number and boolean options
to `0` or `false` respectively when passing `nil` as the value. A value of
`nil` should instead clear the local value, which has different meanings for
different option types:

String options: set the local value to the empty string (`""`) which has the
effect of using the global value of the option for global-local options.

Number and boolean options: set the local value to the global value.

TODO:

- [x] Handle global-local number options (`autoread`, `scrolloff`, `undolevels`)

### CODE REVIEW

**Comment by bfredl** [comment]
Does this work for global-local bool/number options? I e `autoread`, `undolevel` and `scrolloff` should the sentinel value set, rather than copy from global value.

**Comment by gpanders** [comment]
> Does this work for global-local bool/number options? I e `autoread`, `undolevel` and `scrolloff` should the sentinel value set, rather than copy from global value.

Thanks, forgot about these. These will need to be specially handled unfortunately, as they are in `do_set` right now.

File: src/nvim/api/vim.c [hunk 1]
```diff
@@ -696,7 +696,15 @@ Object nvim_get_option_value(String name, Dict(option) *opts, Error *err)
...
     break;
   case 2:
-    rv = BOOLEAN_OBJ(!!numval);
+    switch (numval) {
+      case 0:
+      case 1:
+        rv = BOOLEAN_OBJ(numval);
+        break;
+      default:
+        rv = INTEGER_OBJ(numval);
```
Comment (bfredl) [inline src/nvim/api/vim.c:705]: I'd suggest using `NIL` to represent unset `autoread` value.

Comment (gpanders) [reply src/nvim/api/vim.c:705]: `-1` is used in `do_set`:

```vim
:set autoread
:setl autoread<
:echo &l:autoread
-1
```

Shouldn't we try and keep the API version congruent with the Vim script version?

Comment (bfredl) [reply src/nvim/api/vim.c:705]: what congruence? `&l:autoread` **always** returns an integer value, never a bool, so already using booleans breaks that rule in that case. vim script cannot change it to neither `v:true` / `v:false` nor `v:null` due to backwards compatibility, while I think API functions should use API types whenever possible/reasonable.

Comment (gpanders) [reply src/nvim/api/vim.c]: Ok done.

**Comment by tjdevries** [comment]
yes please, if we can make the API return bools instead of ints, that would be GREAT for the common case of:

```lua
if not vim.opt.xyz:get() then do_something() end
```

where `xyz` always returns 0 or 1.... but that's never falsey in lua.

**Comment by gpanders** [comment]
Ah, this shouldn't be backported since #15996 was not backported.

**Pull request merged**